### PR TITLE
Promise resolution & add a recursive option to all() function

### LINKED ISSUE #57: Promise resolution and reference
Hi everyone,

## TL; DR
Make promise resolution simpler:
```php
$promise = new \GuzzleHttp\Promise\Promise(function () use (&$promise) {
    $promise->resolve('The referenced promise has been resolved');
});
```

to:
```php
$promise = new \GuzzleHttp\Promise\Promise(function (PromiseInterface $promise) {
    $promise->resolve('The promise has been resolved');
});
```

## Example

Actually, when you create a promise from scratch you **must** use a _reference to that promise_ within the wait function, i.e.:

```php
$promise = new \GuzzleHttp\Promise\Promise(function () use (&$promise) {
    $promise->resolve('The referenced promise has been resolved');
});
```

From a DX point of view, this can lead to headaches when it comes to generate promises within a loop. Let's take this as an example:

```php
use GuzzleHttp\Promise\Promise;
use function GuzzleHttp\Promise\all;

require_once __DIR__ . '/vendor/autoload.php';

$translationMap = [
    'banana' => 'banane',
    'apple'  => 'pomme',
];

$fruits = [
    'banana',
];

$promises = [];

foreach ($fruits AS $fruit) {
    $promises[$fruit] = $promise = new Promise(function () use (&$promise, $translationMap, $fruit) {
        $promise->resolve($translationMap[$fruit]);
    });
}

var_dump(all($promises)->wait()); // Expected ['banana' => 'banane'] and indeed got ['banana' => 'banane']
```

In this example, there's only 1 instance of `$promise`, and everything's fine. Now add an `apple` into the `$fruits` array:

```php
use GuzzleHttp\Promise\Promise;
use function GuzzleHttp\Promise\all;

require_once __DIR__ . '/vendor/autoload.php';

$translationMap = [
    'banana' => 'banane',
    'apple'  => 'pomme',
];

$fruits = [
    'banana',
    'apple',
];

$promises = [];

foreach ($fruits AS $fruit) {
    $promises[$fruit] = $promise = new Promise(function () use (&$promise, $translationMap, $fruit) {
        $promise->resolve($translationMap[$fruit]);
    });
}

var_dump(all($promises)->wait()); // Expected ['banana' => 'banane', 'apple' => 'pomme'] 
// But got GuzzleHttp\Promise\RejectionException: Invoking the wait callback did not resolve the promise...
```

Of course, this `RejectionException` would not happen if I chose a different variable name for each `$promise`. But honestly this is ugly.

A simpler solution would be to pass the promise itself directly as an argument into the wait function, instead of relying on a reference, i.e:
```php
$promise = new \GuzzleHttp\Promise\Promise(function (PromiseInterface $promise) {
    $promise->resolve('The promise has been resolved');
});
```

This would require a change into the `Promise::invokeWaitFn()` method:
```php
$wfn($this);
// instead of
$wfn(true);
```

This would simplify a lot the promise resolution.
What's your opinion ?

### PR DESCRIPTION
**Pull request by bpolaszek**
This PR fixes #57 and #46 

For #57 a freshly created promise is injected into its `wait` function for simpler creation / resolution:
```php
use GuzzleHttp\Promise\Promise;
use GuzzleHttp\Promise\PromiseInterface;

$promise = new Promise(function (PromiseInterface $promise) {
    $promise->resolve('The promise has been resolved');
});
```

For #46 the `all()` function, which resolves a stack of promises, has now an option to resolve new promises that were added to the stack during the resolution of the 1st ones.
```php
use GuzzleHttp\Promise\Promise;
use GuzzleHttp\Promise\PromiseInterface;
use function \GuzzleHttp\Promise\all;

$promises = new ArrayIterator();
$promises['Bill'] = new Promise(function (PromiseInterface $promise) use ($promises) {
    $promise->resolve('Clinton');
    $promises['Donald'] = new Promise(function (PromiseInterface $promise) {
        $promise->resolve('Trump');
    });
});

all($promises, true)->then(function ($result) {
    var_dump($result);
})->wait();
```

outputs: 
```php
array (
  "Bill" => "Clinton"
  "Donald" => "Trump"
)
```

Without this only Bill Clinton would have been resolved. 
To avoid BC breaks this option is set to false by default.

**Use case:**
Call an API which have paginated results: for each page, add a new promise to the stack.

### CODE REVIEW

File: src/Promise.php [hunk 1]
```diff
@@ -243,7 +243,7 @@ private function invokeWaitFn()
         try {
             $wfn = $this->waitFn;
             $this->waitFn = null;
-            $wfn(true);
+            $wfn($this);
```
Comment (mtdowling) [inline src/Promise.php]: Why is this change here? What is this meant to do?

File: src/functions.php [hunk 2]
```diff
@@ -230,6 +231,15 @@ function ($reason, $idx, Promise $aggregate) {
     )->then(function () use (&$results) {
         ksort($results);
         return $results;
+    })->then(function ($results) use ($recursive, &$promises) {
```
Comment (mtdowling) [inline src/functions.php]: Can you update the PR so that this extra `then` is only be added if `$recursive` is set to true? That would make this addition have almost zero cost over the current function for the default use case, saving the added overhead of the new promise and function.

File: src/Promise.php (hunk 1 above)
Comment (mtdowling) [inline src/Promise.php]: Ah I see. It's related to https://github.com/guzzle/promises/issues/57. Is this change necessary for this PR? If this isn't needed for this PR, then I'd prefer these PRs to be broken into two different PRs.

I don't see anywhere in the docs that the wait function is expecting a boolean argument or that it is being used anywhere. I think this change and #57 make sense. We should probably update the documentation and docblocks to show this change as well.

Comment (bpolaszek) [inline src/Promise.php]: Hi @mtdowling,

Thank you for reviewing. No pb, I just have to figure out how to make multiple PRs (I'm not a great pull-requester), and I'll update the docs.
I'll look at this next week.

File: src/functions.php (hunk 2 above)
Comment (bpolaszek) [inline src/functions.php]: Good catch, 

I will modify this.

**Comment by bpolaszek** [comment]
Splitted PRs #62 and #63.
